### PR TITLE
add support for formatted strings (fixes #130)

### LINKED ISSUE #130: strip IRC colors (bold, underline, etc.)
I would like to implement a utility function that strips IRC colors. 
Where would you suggest this to be implemented?

### PR DESCRIPTION
**Pull request by freddyb**
I'm relatively new to rust<sup>[1](#)</sup> and this is my first Trait implementation. Maybe like this?

---

<sup>[1](#)</sup> I've done something like this in other languages before and I sorta know IRC. So, I hope this isn't too embarrassing :)

### CODE REVIEW

**Comment by aatxe** [comment]
Thanks for the PR, I'll go through and leave some comments. 😄

File: src/proto/colors.rs
Comment (aatxe) [inline src/proto/colors.rs]: Given that this is an extension trait for all strings, I'd prefer for this to be named `FormattedStringExt` to reflect that. I think having `Ext` in the name is common, but if nothing else, it's good for consistency with `ChannelExt` here.

File: src/proto/colors.rs
Comment (aatxe) [inline src/proto/colors.rs]: I think of these constants as being implementation details rather than part of the interface. Do you think think there's a reason to expose them publcily? If not, I would move them outside of the trait definition.

File: src/proto/colors.rs
Comment (aatxe) [inline src/proto/colors.rs]: I think names like `strip_formatting` or `strip_irc_formatting` better reflects the effect on the string (given bold, underline, italics).

File: src/proto/mod.rs
Comment (aatxe) [inline src/proto/mod.rs]: You'll need to add `pub mod colors` above for the Rust compiler to compile the module you wrote. This is why the Travis build failed.

File: src/proto/colors.rs
Comment (aatxe) [inline src/proto/colors.rs]: My brain type checker could be failing me, but I think the return type here is actually `Cow<str>`. So, you should be able to `use std::borrow::Cow` and change the return type in the impls and the trait definition to `Cow<str>`.

**Comment by freddyb** [comment]
(making some progress, found some errors and fixing locally)

**Review by aatxe — CHANGES_REQUESTED** [review]
Thanks again for the PR, it's very appreciated! 😄 I've left a few suggestions about names, and some comments that might help with the errors you've found. Let me know if you need any help though (or you can ping me on Mozilla IRC as `awe`)! 😄

**Comment by freddyb** [comment]
Thanks for the speedy review. I've got a bit further.

But I can't find a way to create the regex early on and just keep it around.
Now I wonder where I'd want to create it to avoid recompilation :thinking: 

```
error[E0015]: calls in constants are limited to struct and enum constructors
  --> src/proto/colors.rs:10:29
   |
10 | const STRIP_COLORS: Regex = Regex::new(r"\x1F|\x02|\x12|\x0F|\x16|\x03(?:\d{1,2}(?:,\d{1,2})?)?").unwrap();
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by aatxe** [comment]
There's a few ways forward that I see:

1. Just create the regex every time. I'm not sure how expensive it is to compile, but it's a pretty straightforward regex, so maybe not too bad. The documentation for regex suggests maybe a few microseconds for "small" regexps.
2. We can use the [`lazy_static`](https://crates.io/crates/lazy_static) crate to lazily initialize it. The downside is that this is another dependency.
3. Instead of a trait, we could have some struct like `IrcFormatting` that could keep the regex as internal state. This would be a rather different API though because you'd have to construct an instance of it in advance and use it as opposed to just write `some_normal_string.strip_irc_formatting()`.
4. Hand-write the stripping procedure without `regex`. This is a bit harder to get right because of the way the colors work (I have a probably buggy implementation for parsing IRC formatting in [`alectro`](https://github.com/aatxe/alectro), you can find it split between [here](https://github.com/aatxe/alectro/blob/master/src/view/widget/chatbuf.rs#L77) and [here](https://github.com/aatxe/alectro/blob/master/src/view/widget/chatbuf.rs#L156)), but it doesn't require any extra dependencies.

I think I'm inclined to say 2 or 4 make the most sense. If this requires two dependencies (`regex` and `lazy_static`), I'll probably end up making it optional via a feature, but that's alright. I think some time soon we're supposed to be able to do compile-time compilation for regexps and then we might even be able to get rid of `lazy_static` at that point.

**Comment by freddyb** [comment]
I'll try 4, i.e., implementing this without regexes.
Without having looked at your implementation, my naive approach would be to go through all characters and drop those for normal,underline,bold,reverse. I'm not sure how to exactly to look ahead to see how much more I'm supposed to drop, when I encounter the color character. Maybe I could peek() and then mark them for deletion in a following iteration.

I'm sure I'll figure something out, once I'm back at a computer.

**Comment by freddyb** [comment]
I've rewritten this into its own repo, to have something that's easier tested in its own, the tests pass, butbut I'm not happy with the solution at all :| Maybe you want to take a look anyway?

https://github.com/freddyb/leap

**Comment by aatxe** [comment]
Hi @freddyb, sorry for the delay. Your solution in leap looks great! My only suggestions are really stylistic: it's more idiomatic to use match in this context (with pattern guards for the extra conditions), and also to not use `return`. You might also find it helpful to run [clippy](https://github.com/rust-lang-nursery/rust-clippy) (which I _think_ should suggest both of those things and maybe others 😄). You might find it a bit easier to run clippy and make those changes in your little leap repo though because we already have some lints that need to be fixed in the irc crate itself and will add noise.

**Comment by freddyb** [comment]
Thanks for the pointer. I've rewritten this using match statements and I'm much happier with the readability.
It still lives in https://github.com/freddyb/leap/blob/master/src/lib.rs but I'll update my pull request to use that function in the following days.

**Comment by aatxe** [comment]
Sounds good! Looking forward to it! 😄

**Comment by aatxe** [comment]
Oh, I forgot one other small thing! You can replace the test on line 30 (negated ||'s) with `!['\x02', '\x1F', '\x16', '\x0F'].contains(cur)` which is a bit clearer.

**Comment by freddyb** [comment]
sorry for taking so long @aatxe, but it works now. Can you review again?

File: Cargo.toml
Comment (aatxe) [inline Cargo.toml]: Can you remove the regex dependency now that you've rewritten the implementation without it?

File: src/proto/colors.rs
Comment (aatxe) [inline src/proto/colors.rs]: Can you also take these clippy attributes off? Since clippy-as-a-compiler-plugin doesn't work on stable, it's typical to just have a nightly install of clippy and just do `cargo +nightly clippy` instead.

**Review by aatxe — COMMENTED** [review]
Overall, the changes look good! There's a couple tiny things to do to clean up a bit. I left two small comments, but a bigger thing is that it'd be great if you could throw the tests you had in your repository into a test submodule (with `#[cfg(test)]`). Wouldn't want any regressions in the future! But I can add tests myself after merging, if you'd prefer. Besides that, I think a git squash is in order. Then, we're ready to go! 😄

File: Cargo.toml
Comment (freddyb) [reply Cargo.toml]: oops :) done

File: src/proto/colors.rs
Comment (freddyb) [reply src/proto/colors.rs]: :+1:

**Comment by freddyb** [comment]
I've addressed your smaller comments. Currently trying to figure out how to create a test submodule and where to place it. But feel free to merge right now and file another issue for that, if you want.

**Comment by freddyb** [comment]
Also, I've squashed commits :)

**Comment by aatxe** [comment]
You should be able to put something like:

```rust
#[cfg(test)]
mod test {
...
}
```

at the end of `colors.rs` with all of the tests in it. You can see this pattern a lot throughout the irc crate, but [here's a good example for Messages](https://github.com/aatxe/irc/blob/develop/src/proto/message.rs#L278).

**Review by aatxe — APPROVED** [review]

**Comment by freddyb** [comment]
now with tests :heavy_check_mark:

**Comment by aatxe** [comment]
Looks great! Thanks, @freddyb! 😄

**Pull request merged**
